### PR TITLE
Core/Creature: UNIT_FLAG_IMMUNE_TO_NPC check only vs non-player targets

### LINKED ISSUE #14744: [3.3.5] Sunstrider guardian aren't attacking alliance player
Sunstrider guardian aren't  attacking alliance player

schermata 2015-05-22 alle 22 40 30

tried at sun strider isle bridge. manny npc there aren't attacking me. Instead Silvermoon guards seem ok.

for example this npc ranger sirloin aren't attacking neither
http://www.wowhead.com/npc=15940/ranger-selron

DB 335.58
rev 0148832 2015-05-21

### PR DESCRIPTION
**Pull request by mik1893**
closes #14744

### CODE REVIEW

File: src/server/game/Entities/Creature/Creature.cpp [hunk 1]
```diff
@@ -1370,7 +1370,10 @@ bool Creature::CanStartAttack(Unit const* who, bool force) const
     if (IsCivilian())
         return false;
 
-    if (HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_IMMUNE_TO_NPC))
+    if (HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_IMMUNE_TO_NPC) && who->GetTypeId() != TYPEID_PLAYER)  // flag is valid only for non player characters
+        return false;
+
+    if (HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_IMMUNE_TO_PC) && who->GetTypeId() == TYPEID_PLAYER)  // immune to PC and target is a player, return false
         return false;
```
Comment (joschiwald) [inline src/server/game/Entities/Creature/Creature.cpp]: `UNIT_FLAG_IMMUNE_TO_(N)PC` checks are done in `Creature::CanCreatureAttack`
we could delete this check here?

**Comment by mik1893** [comment]
let me check, i guess it's gonna be the same. In any case with the check that was done here, things were not working. I agree those checks should be all stay in the same place.

File: src/server/game/Entities/Creature/Creature.cpp (hunk 1 above)
Comment (tkrokli) [inline src/server/game/Entities/Creature/Creature.cpp]: I just tested your theory by removing the existing lines 1373 + 1374 (red and white lines above) and the Sunstrider Guardians react correctly by attacking in that compile. They continued to attack my hunter character's pet after switching to GM ON, though. I thought that a player combat pet would follow the player character's GM state.

**Comment by ariel-** [comment]
This affects player controled creatures (pets), now can be attacked by NPCs supposed to be static, like http://www.wowhead.com/npc=38757

**Comment by sirikfoll** [comment]
Confirmed, 
Defile(http://www.wowhead.com/npc=38757) follow Hunter's pet/Shaman's totem

**Comment by Runico** [comment]
Yeap, Defile now can attacks :dancer:

**Comment by ariel-** [comment]
Is it considered hackfix to put extra_flag_trigger (128) on this npc?

**Comment by Carbenium** [comment]
Summoning @mik1893 : Could you take a look what the patch broke?

File: src/server/game/Entities/Creature/Creature.cpp [hunk 2]
```diff
@@ -1370,9 +1370,6 @@ bool Creature::CanStartAttack(Unit const* who, bool force) const
     if (IsCivilian())
         return false;
 
-    if (HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_IMMUNE_TO_NPC))
```
Comment (Nyeriah) [inline src/server/game/Entities/Creature/Creature.cpp]: what about this?

```cpp
if (who->GetTypeId == TYPEID_UNIT && HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_IMMUNE_TO_NPC))
```

**Comment by tkrokli** [comment]
> **ariel-** commented 4. June 2015, 21:02 CEST
This affects player controled creatures (pets), now can be attacked by NPCs supposed to be static, like http://www.wowhead.com/npc=38757

***
> **sirikfoll** commented 4. June 2015, 21:02 CEST
Confirmed,
Defile(http://www.wowhead.com/npc=38757) follow Hunter's pet/Shaman's totem
***

Is this somehow related to what I commented in https://github.com/TrinityCore/TrinityCore/issues/14744#issuecomment-106165356 about NPC attacking a hunter's pet even if the player turns GM on? It happened after applying @joschiwald's suggested fix used in this PR. This PR fixed the issue with PvP NPCs not attacking players of the opposite faction, though.

**Comment by mik1893** [comment]
this is simply because there is no check about pets and NPC flag in Unit::_IsValidAttackTarget.
I will add the check here, so we still stay in the same function with those checks

**Comment by Killyana** [comment]
@mik1893 Your fix works for me.

**Pull request merged**